plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/parameterInfo/KotlinHighLevelFunctionParameterInfoHandler.kt

```diff
@@ -0,0 +1,451 @@
+
+                // TODO: When resolvedCall is KtFunctionalTypeVariableCall, the candidate is FunctionN.invoke() and parameter names are "p1", "p2", etc.
+                // We need to get the type of the target variable, and retrieve the parameter names from the type (KtFunctionalType).
+                // The names need to be added to KtFunctionalType (currently only types are there) and populated in KtSymbolByFirBuilder.TypeBuilder.
+
+                val parameterToText = valueParameters.associateWith { parameter ->
+                    // TODO: Add hasSynthesizedParameterNames to HL API.
+                    // See resolveValueParameters() in core/descriptors.jvm/src/org/jetbrains/kotlin/load/java/lazy/descriptors/LazyJavaScope.kt
+                    val hasSynthesizedParameterNames = false
+                    renderParameter(parameter, includeName = !hasSynthesizedParameterNames)
+                }
+
+                // TODO: The argument mapping should also be per-candidate once we have all candidates available.
+                val argumentMapping = resolvedCall.argumentMapping
+
+                // TODO: This should be changed when there are multiple candidates available; need to know which one the call is resolved to
+                val isCallResolvedToCandidate = candidates.size == 1
+
+                candidateInfo.callInfo = CallInfo(
+                    callElement,
+                    valueArguments,
+                    arguments,
+                    argumentMapping,
+                    valueParameters,
+                    parameterToText,
+                    isCallResolvedToCandidate
+                )
+            }
+        }
+    }
+
+    private fun areEqual(first: KtFunctionLikeSymbol, second: KtFunctionLikeSymbol): Boolean {
+        // TODO: Should we do signature matching instead?
+        val firstPsi = first.psi ?: return false
+        val secondPsi = second.psi ?: return false
+        return firstPsi.manager.areElementsEquivalent(firstPsi, secondPsi)
+    }
+
+    protected open fun getParameterIndex(context: UpdateParameterInfoContext, argumentList: TArgumentList): Int {
+        val offset = context.offset
+        return argumentList.allChildren
+            .takeWhile { it.startOffset < offset }
+            .count { it.node.elementType == KtTokens.COMMA }
+    }
+
+    private fun KtAnalysisSession.renderParameter(parameter: KtValueParameterSymbol, includeName: Boolean): String {
+        return buildString {
+            val annotationFqNames =
+                parameter.annotations
+                    .filter {
+                        // For primary constructor parameters, the annotation use site must be "param" or unspecified.
+                        it.useSiteTarget == null || it.useSiteTarget == AnnotationUseSiteTarget.CONSTRUCTOR_PARAMETER
+                    }
+                    .mapNotNull { it.classId?.asSingleFqName() }
```

Comment:
It looks like FE1.0 does not print package names for annotations. Maybe doing the same?

Reply:
I am only printing the short name. See call to `it.shortName()` a few lines below. I'll modify one of the tests `annotations/FunctionCall.kt` to put it in a package to test that only the short name is rendered.

This part is only for filtering out nullability annotations, so we need the FQ name.